src/postprocess/refine.jl

```diff
@@ -0,0 +1,159 @@
+# Refinement of some quantities of interest (density, forces) following the
+# strategy described in [CDKL2021].
+#
+# [CDKL2021]:
+#     E. Cancès, G. Dusson, G. Kemlin, and A. Levitt
+#     *Practical error bounds for properties in plane-wave electronic structure
+#     calculations* Preprint, 2021. [arXiv](https://arxiv.org/abs/2111.01470)
+
+"""
+Result of calling the [`refine_scfres`](@ref) function.
+- `basis`: Refinement basis, larger than the basis used to
+           run a first [`self_consistent_field`](@ref) computation.
+- `ψ`, `ρ`, `occupation`: Quantities from the scfres, transferred to the refinement basis
+                          and with virtual orbitals removed.
+- `δψ`, `δρ`: First order corrections to the wavefunctions and density.
+              The sign is such that the refined quantities are ψ - δψ and ρ - δρ.
+"""
+struct RefinementResult
+    basis::PlaneWaveBasis
+    ψ
+    ρ
+    occupation
+    δψ
+    δρ
+end
+
+"""
+Transfer the result of an SCF to a larger basis set,
+and compute first order corrections ("refinements") to the wavefunctions and density.
+
+Only full occupations are currently supported.
+
+Returns a [`RefinementResult`](@ref) instance that can be used to refine quantities of interest,
+through [`refine_density`](@ref) and [`refine_forces`](@ref).
+"""
+function refine_scfres(scfres, basis_ref::PlaneWaveBasis{T}; ΩpK_tol,
+                       occ_threshold=default_occupation_threshold(T), kwargs...) where {T}
+    basis = scfres.basis
+
+    @assert basis.model.lattice == basis_ref.model.lattice
+    @assert length(basis.kpoints) == length(basis_ref.kpoints)
+    @assert all(basis.kpoints[ik].coordinate == basis_ref.kpoints[ik].coordinate
+                for ik in 1:length(basis.kpoints))
+
+    # Virtual orbitals must be removed
+    ψ, occ = select_occupied_orbitals(basis, scfres.ψ, scfres.occupation; threshold=occ_threshold)
+    check_full_occupation(basis, occ)
+
+    ψr = transfer_blochwave(ψ, basis, basis_ref)
+    ρr = transfer_density(scfres.ρ, basis, basis_ref)
+    _, hamr = energy_hamiltonian(basis_ref, ψr, occ; ρ=ρr)
+
+    # Compute the residual R(P)
+    # This recomputes ρr internally, but it's only a minor inefficiency.
+    res = compute_projected_gradient(basis_ref, ψr, occ)
+
+    # Compute the projection of the residual onto the high and low frequencies
+    resLF = transfer_blochwave(res, basis_ref, basis)
+    resHF = res - transfer_blochwave(resLF, basis, basis_ref)
+
+    """
+    Apply M^{-1}.
+
+    M applied to the i-th band is defined by
+      M_i = P^⟂ T_i^{1/2} P^⟂ T_i^{1/2} P^⟂,
+    with the diagonal operator
+      T_i = kinetic energy + mean kinetic energy of i-th band.
+
+    To invert M_i, use P^⟂ T_i^{-1} P^⟂ as the preconditioner.
+
+    See [CDKL2021] for more details.
+    """
```

Comment:
Put as comments. Also for longer comments like this we prefer to have them *outside* the actual function to ensure that the actual code is condensed (one gets a better overview what's going on in this way).

Reply:
I think that this should be kept as close as possible to where `M` is actually defined because there are many things going on. An alternative is to move `M` out but I don't think it's relevant in other places.

Reply:
Moved to a comment at the top of the file.